File: src/git.js
import { exec, execFile, spawn } from 'child_process';
import { promisify } from 'util';
import { logMessage } from './logger.js';

const execAsync = promisify(exec);
const execFileAsync = promisify(execFile);

/**
 * Executes a Git command in the current working directory.
 * @param {string|string[]} args - The arguments or array of arguments to pass to the git CLI.
 * @returns {Promise<{success: boolean, stdout: string, stderr: string, error?: string}>}
 */
export async function runGitCommand(args, cwd = process.cwd()) {
  const cmdStr = Array.isArray(args) ? `git ${args.join(' ')}` : `git ${args}`;
  try {
    let stdout, stderr;
    if (Array.isArray(args)) {
      const { stdout: out, stderr: err } = await execFileAsync('git', args, { cwd });
      stdout = out;
      stderr = err;
    } else {
      const { stdout: out, stderr: err } = await execAsync(`git ${args}`, { cwd });
      stdout = out;
      stderr = err;
    }
    logMessage(`Executed: ${cmdStr}`, 'INFO');
    return { success: true, stdout: stdout.trim(), stderr: stderr.trim() };
  } catch (error) {
    const errorMsg = error.message;
    const stderrMsg = error.stderr?.trim() || '';
    logMessage(`Failed: ${cmdStr} - Error: ${errorMsg} - Stderr: ${stderrMsg}`, 'ERROR');
    return { 
      success: false, 
      error: errorMsg, 
      stdout: error.stdout?.trim() || '', 
      stderr: stderrMsg
    };
  }
}

/**
 * Checks if the current directory is a git repository.
 * @returns {Promise<boolean>}
 */
export async function isGitRepository() {
  const res = await runGitCommand('rev-parse --is-inside-work-tree');
  return res.success && res.stdout === 'true';
}

/**
 * Returns a list of changed, untracked, and deleted files.
 * @returns {Promise<Array<{path: string, type: string, state: 'staged'|'unstaged'|'both'|'conflict', rawStatus: string}>>}
 */
export async function getChangedFiles() {
  const res = await runGitCommand('status --porcelain');
  if (!res.success || !res.stdout) return [];
  
  return res.stdout.split('\n').filter(Boolean).map(line => {
    const status = line.slice(0, 2);
    const path = line.slice(3).replace(/^"|"$/g, ''); // strip optional quotes around spaces/unicode
    let state = 'unstaged';
    let type = 'Modified';
    
    if (status === '??') {
      type = 'Untracked';
    } else if (status === ' A' || status === 'A ') {
      type = 'Added';
      state = status === 'A ' ? 'staged' : 'unstaged';
    } else if (status === ' D' || status === 'D ') {
      type = 'Deleted';
      state = status === 'D ' ? 'staged' : 'unstaged';
    } else if (status === 'M ') {
      type = 'Modified';
      state = 'staged';
    } else if (status === ' M') {
      type = 'Modified';
      state = 'unstaged';
    } else if (status === 'MM') {
      type = 'Modified (Staged & Unstaged)';
      state = 'both';
    } else if (status === 'R ') {
      type = 'Renamed';
      state = 'staged';
    } else if (status === 'UU' || status === 'U' || status.includes('U')) {
      type = 'Conflict';
      state = 'conflict';
    }
    
    return { path, type, state, rawStatus: status };
  });
}

export async function stageFiles(files) {
  if (files === 'all') {
    return runGitCommand(['add', '.']);
  }
  return runGitCommand(['add', ...files]);
}

/**
 * Unstages files (reset HEAD).
 * @param {string[]|'all'} files 
 */
export async function unstageFiles(files) {
  if (files === 'all') {
    return runGitCommand(['reset', 'HEAD']);
  }
  return runGitCommand(['reset', 'HEAD', ...files]);
}

/**
 * Commits currently staged changes. Uses spawn with stdio inherit to support GPG signing.
 * @param {string} message 
 */
export async function commit(message) {
  return new Promise((resolve) => {
    if (process.stdin.isTTY) {
      process.stdin.pause();
    }
    const child = spawn('git', ['commit', '-m', message], { stdio: 'inherit' });
    child.on('close', (code) => {
      if (process.stdin.isTTY) {
        process.stdin.resume();
      }
      if (code === 0) {
        resolve({ success: true, stdout: 'Commit created successfully' });
      } else {
        resolve({ success: false, error: `Git commit exited with code ${code}` });
      }
    });
  });
}

/**
 * Retrieves the local branches.
 * @returns {Promise<Array<{name: string, current: boolean}>>}
 */
export async function getLocalBranches() {
  const res = await runGitCommand('branch');
  if (!res.success || !res.stdout) return [];
  return res.stdout.split('\n').filter(Boolean).map(line => {
    const name = line.replace(/^\*\s+/, '').trim();
    const current = line.startsWith('*');
    return { name, current };
  });
}

/**
 * Gets the current branch and its ahead/behind counts against upstream.
 * @returns {Promise<{branch: string, ahead: number, behind: number, hasUpstream: boolean}>}
 */
export async function getAheadBehind() {
  let branchRes = await runGitCommand('symbolic-ref --short HEAD');
  if (!branchRes.success) {
    branchRes = await runGitCommand('rev-parse --abbrev-ref HEAD');
  }
  if (!branchRes.success) return { branch: 'unknown', ahead: 0, behind: 0, hasUpstream: false };
  const branch = branchRes.stdout;
  
  const upstreamRes = await runGitCommand('rev-parse --abbrev-ref HEAD@{upstream}');
  if (!upstreamRes.success) {
    return { branch, ahead: 0, behind: 0, hasUpstream: false };
  }
  
  const diffRes = await runGitCommand('rev-list --left-right --count HEAD...HEAD@{upstream}');
  if (!diffRes.success) {
    return { branch, ahead: 0, behind: 0, hasUpstream: true };
  }
  
  const [ahead, behind] = diffRes.stdout.split(/\s+/).map(Number);
  return { branch, ahead, behind, hasUpstream: true };
}

/**
 * Pulls changes from remote.
 */
export async function pull() {
  return runGitCommand('pull --rebase');
}

/**
 * Pushes changes to remote.
 */
export async function push() {
  const aheadBehind = await getAheadBehind();
  if (!aheadBehind.hasUpstream) {
    // If no upstream is set, push and set upstream to origin current-branch
    return runGitCommand(`push --set-upstream origin ${aheadBehind.branch}`);
  }
  return runGitCommand('push');
}

/**
 * List stashes.
 * @returns {Promise<string[]>}
 */
export async function getStashes() {
  const res = await runGitCommand('stash list');
  if (!res.success || !res.stdout) return [];
  return res.stdout.split('\n').filter(Boolean);
}

export async function stashSave(message) {
  const args = ['stash', 'push'];
  if (message) {
    args.push('-m', message);
  }
  return runGitCommand(args);
}

/**
 * Pops the last stashed changes.
 */
export async function stashPop() {
  return runGitCommand('stash pop');
}

/**
 * Gets the recent commits.
 * @param {number} limit 
 * @returns {Promise<Array<{hash: string, message: string, author: string}>>}
 */
export async function getRecentCommits(limit = 3) {
  const res = await runGitCommand(`log -n ${limit} --pretty=format:"%h|%s|%an"`);
  if (!res.success || !res.stdout) return [];
  
  return res.stdout.split('\n').filter(Boolean).map(line => {
    const [hash, message, author] = line.split('|');
    return { hash, message, author };
  });
}

File: internal/git/git.go
package git

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"

	"github.com/dwaipayanray95/devD/internal/logger"
)

type GitResult struct {
	Success bool
	Stdout  string
	Stderr  string
	Error   error
}

type ChangedFile struct {
	Path      string
	Type      string
	State     string // "staged", "unstaged", "both", "conflict"
	RawStatus string
}

type AheadBehindResult struct {
	Branch      string
	Ahead       int
	Behind      int
	HasUpstream bool
}

type CommitInfo struct {
	Hash    string
	Message string
	Author  string
}

type BranchInfo struct {
	Name      string
	IsCurrent bool
	IsRemote  bool
}

func RunGitCommand(args []string) GitResult {
	cmdStr := "git " + strings.Join(args, " ")
	cmd := exec.Command("git", args...)
	
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	stdoutStr := strings.TrimSpace(stdout.String())
	stderrStr := strings.TrimSpace(stderr.String())

	if err != nil {
		logger.LogMessage(fmt.Sprintf("Failed: %s - Error: %s - Stderr: %s", cmdStr, err.Error(), stderrStr), "ERROR")
		return GitResult{
			Success: false,
			Stdout:  stdoutStr,
			Stderr:  stderrStr,
			Error:   err,
		}
	}

	logger.LogMessage(fmt.Sprintf("Executed: %s", cmdStr), "INFO")
	return GitResult{
		Success: true,
		Stdout:  stdoutStr,
		Stderr:  stderrStr,
	}
}

func IsGitRepository() bool {
	res := RunGitCommand([]string{"rev-parse", "--is-inside-work-tree"})
	return res.Success && res.Stdout == "true"
}

func GetChangedFiles() ([]ChangedFile, error) {
	res := RunGitCommand([]string{"status", "--porcelain"})
	if !res.Success {
		return nil, res.Error
	}
	if res.Stdout == "" {
		return nil, nil
	}

	var files []ChangedFile
	lines := strings.Split(res.Stdout, "\n")
	for _, line := range lines {
		if len(line) < 4 {
			continue
		}
		status := line[:2]
		filePath := strings.Trim(line[3:], "\" ")
		
		state := "unstaged"
		fileType := "Modified"

		if status == "??" {
			fileType = "Untracked"
		} else if status == " A" || status == "A " {
			fileType = "Added"
			if status == "A " {
				state = "staged"
			}
		} else if status == " D" || status == "D " {
			fileType = "Deleted"
			if status == "D " {
				state = "staged"
			}
		} else if status == "M " {
			fileType = "Modified"
			state = "staged"
		} else if status == " M" {
			fileType = "Modified"
			state = "unstaged"
		} else if status == "MM" {
			fileType = "Modified (Staged & Unstaged)"
			state = "both"
		} else if status == "R " {
			fileType = "Renamed"
			state = "staged"
		} else if status == "UU" || status == "U" || strings.Contains(status, "U") {
			fileType = "Conflict"
			state = "conflict"
		}

		files = append(files, ChangedFile{
			Path:      filePath,
			Type:      fileType,
			State:     state,
			RawStatus: status,
		})
	}
	return files, nil
}

func StageFiles(files []string) GitResult {
	if len(files) == 1 && files[0] == "all" {
		return RunGitCommand([]string{"add", "."})
	}
	args := append([]string{"add"}, files...)
	return RunGitCommand(args)
}

func UnstageFiles(files []string) GitResult {
	if len(files) == 1 && files[0] == "all" {
		return RunGitCommand([]string{"reset", "HEAD"})
	}
	args := append([]string{"reset", "HEAD"}, files...)
	return RunGitCommand(args)
}

func Commit(message string) GitResult {
	cmd := exec.Command("git", "commit", "-m", message)
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	err := cmd.Run()
	if err != nil {
		return GitResult{
			Success: false,
			Error:   err,
		}
	}
	return GitResult{
		Success: true,
		Stdout:  "Commit created successfully",
	}
}

func GetLocalBranches() ([]string, error) {
	res := RunGitCommand([]string{"branch"})
	if !res.Success {
		return nil, res.Error
	}
	var branches []string
	lines := strings.Split(res.Stdout, "\n")
	for _, line := range lines {
		name := strings.TrimSpace(strings.ReplaceAll(line, "*", ""))
		if name != "" {
			branches = append(branches, name)
		}
	}
	return branches, nil
}

func GetAheadBehind() (AheadBehindResult, error) {
	branchRes := RunGitCommand([]string{"symbolic-ref", "--short", "HEAD"})
	if !branchRes.Success {
		branchRes = RunGitCommand([]string{"rev-parse", "--abbrev-ref", "HEAD"})
	}
	if !branchRes.Success {
		return AheadBehindResult{Branch: "unknown", Ahead: 0, Behind: 0, HasUpstream: false}, errors.New("cannot resolve HEAD branch")
	}
	branch := branchRes.Stdout

	upstreamRes := RunGitCommand([]string{"rev-parse", "--abbrev-ref", "HEAD@{upstream}"})
	if !upstreamRes.Success {
		return AheadBehindResult{Branch: branch, Ahead: 0, Behind: 0, HasUpstream: false}, nil
	}

	diffRes := RunGitCommand([]string{"rev-list", "--left-right", "--count", "HEAD...HEAD@{upstream}"})
	if !diffRes.Success {
		return AheadBehindResult{Branch: branch, Ahead: 0, Behind: 0, HasUpstream: true}, nil
	}

	parts := strings.Fields(diffRes.Stdout)
	if len(parts) < 2 {
		return AheadBehindResult{Branch: branch, Ahead: 0, Behind: 0, HasUpstream: true}, nil
	}

	ahead, _ := strconv.Atoi(parts[0])
	behind, _ := strconv.Atoi(parts[1])

	return AheadBehindResult{
		Branch:      branch,
		Ahead:       ahead,
		Behind:      behind,
		HasUpstream: true,
	}, nil
}

func Pull() GitResult {
	return RunGitCommand([]string{"pull", "--rebase"})
}

func Push() GitResult {
	ab, err := GetAheadBehind()
	if err == nil && !ab.HasUpstream {
		return RunGitCommand([]string{"push", "--set-upstream", "origin", ab.Branch})
	}
	return RunGitCommand([]string{"push"})
}

func GetStashes() ([]string, error) {
	res := RunGitCommand([]string{"stash", "list"})
	if !res.Success {
		return nil, res.Error
	}
	if res.Stdout == "" {
		return nil, nil
	}
	return strings.Split(res.Stdout, "\n"), nil
}

func StashSave(message string) GitResult {
	args := []string{"stash", "push"}
	if message != "" {
		args = append(args, "-m", message)
	}
	return RunGitCommand(args)
}

func StashPop() GitResult {
	return RunGitCommand([]string{"stash", "pop"})
}

func GetRecentCommits(limit int) ([]CommitInfo, error) {
	res := RunGitCommand([]string{"log", fmt.Sprintf("-n %d", limit), "--pretty=format:%h|%s|%an"})
	if !res.Success {
		return nil, res.Error
	}
	if res.Stdout == "" {
		return nil, nil
	}

	var commits []CommitInfo
	lines := strings.Split(res.Stdout, "\n")
	for _, line := range lines {
		parts := strings.Split(line, "|")
		if len(parts) >= 3 {
			commits = append(commits, CommitInfo{
				Hash:    parts[0],
				Message: parts[1],
				Author:  parts[2],
			})
		}
	}
	return commits, nil
}
